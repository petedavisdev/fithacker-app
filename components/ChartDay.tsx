import { Text, View } from 'react-native';
import { useTranslation } from 'react-i18next';
import { Activity, ActivityDay } from '../constants/ACTIVITIES';
import { Link } from 'expo-router';
import { getDateInfo } from '../utils/dateInfo';

type ChartDayProps = {
	date: string;
	activities: ActivityDay;
	filter?: Activity;
};

export function ChartDay(props: ChartDayProps) {
	const { t } = useTranslation();
	const dateInfo = getDateInfo(props.date);

	const DATE_TEXT_COLORS = {
		future: 'text-slate-500',
		today: 'text-pink-400',
		tomorrow: 'text-slate-500',
		weekend: 'text-yellow-500',
		weekday: 'text-cyan-500',
	};

	const dateTextColor = DATE_TEXT_COLORS[dateInfo.category];

	const DATE_LINE_COLORS = {
		future: 'bg-black',
		tomorrow: 'bg-slate-700 shadow shadow-slate-700',
		today: 'bg-pink-500 shadow shadow-pink-500',
		weekend: 'bg-yellow-500 shadow shadow-yellow-500',
		weekday: 'bg-cyan-500 shadow shadow-cyan-500',
	};

	const dateLineColor = DATE_LINE_COLORS[dateInfo.category];

	return (
		<Link
			key={props.date}
			href={`/${props.date}`}
			disabled={dateInfo.category === 'future'}
		>
			<View className="justify-end items-center h-96 gap-2">
				{props.activities.map((activityItem, index) => {
					const note =
						typeof activityItem !== 'string' && activityItem[1];
					const activity = note ? activityItem[0] : activityItem;

					if (!props.filter || activity === props.filter)
						return (
							<View key={index} className="relative">
								{props.filter && note && (
									<View className="absolute -top-12 w-full -rotate-90">
										<Text
											key={`note${index}`}
											className="font-mono text-pink-500 w-80 h-12 p-3"
										>
											{note}
										</Text>
									</View>
								)}
								<Text
									key={`activity${index}`}
									className="text-yellow-500 text-4xl"
								>
									{typeof activity === 'string'
										? activity
										: activity[0]}
								</Text>
							</View>
						);
				})}

				<View className={`h-[2px] w-12 ${dateLineColor}`} />

				<Text className={`font-mono leading ${dateTextColor}`}>
					{t(`_day.${dateInfo.dayIndex}`).slice(0, 3)}
				</Text>
			</View>
		</Link>
	);
}
