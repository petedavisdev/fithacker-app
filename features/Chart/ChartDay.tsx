import { Text, View } from 'react-native';
import { useTranslation } from 'react-i18next';
import type { Exercise, ExerciseDay } from '../EXERCISES';
import { type Href, Link, useLocalSearchParams } from 'expo-router';
import { getDateInfo } from '../dateInfo';

type ChartDayProps = {
	date: string;
	exercises?: ExerciseDay | ['ðŸ”’'];
};

export function ChartDay(props: ChartDayProps) {
	const { t } = useTranslation();
	const { filter } = useLocalSearchParams<{ filter?: Exercise | '' }>();
	const dateInfo = getDateInfo(props.date);
	const isLocked = props.exercises?.[0] === 'ðŸ”’';
	const isDisabled = isLocked || dateInfo.category === 'future';

	const DATE_TEXT_COLORS = {
		future: 'text-slate-500',
		today: 'text-pink-400',
		tomorrow: 'text-slate-400',
		weekend: 'text-yellow-500',
		weekday: 'text-cyan-500',
	};

	const dateTextColor = DATE_TEXT_COLORS[dateInfo.category];

	const DATE_LINE_COLORS = {
		future: 'bg-black',
		tomorrow: 'bg-slate-700 shadow shadow-slate-700',
		today: 'bg-pink-500 shadow shadow-pink-500',
		weekend: 'bg-yellow-500 shadow shadow-yellow-500',
		weekday: 'bg-cyan-500 shadow shadow-cyan-500',
	};

	const dateLineColor = DATE_LINE_COLORS[dateInfo.category];

	const DATE_UNDERLINE_COLORS = {
		future: '',
		tomorrow: 'border-slate-600',
		today: 'border-pink-600',
		weekend: 'border-yellow-700',
		weekday: 'border-cyan-700',
	};

	const dateUnderlineColor = DATE_UNDERLINE_COLORS[dateInfo.category];

	return (
		<Link href={`/?date=${props.date}` as Href} disabled={isDisabled}>
			<View className="justify-end items-center h-96 gap-2">
				{props.exercises?.map((exerciseItem, index) => {
					const note =
						typeof exerciseItem !== 'string' && exerciseItem[1];
					const exercise = note ? exerciseItem[0] : exerciseItem;

					return (
						<View
							key={`${props.date}${index}`}
							className="relative"
						>
							{!!filter && note && (
								<View className="absolute -top-12 w-full -rotate-90">
									<Text className="font-mono text-pink-500 w-80 h-12 p-3">
										{note}
									</Text>
								</View>
							)}

							<Text className="text-yellow-500 text-4xl">
								{typeof exercise === 'string'
									? exercise
									: exercise[0]}
							</Text>
						</View>
					);
				})}

				<View className={`h-[2px] w-12 ${dateLineColor}`} />

				<Text
					className={`font-mono leading pb-0.5 border-b ${dateTextColor} ${
						isDisabled
							? 'border-transparent'
							: `${dateUnderlineColor}`
					}`}
				>
					{t(`_day.${dateInfo.dayIndex}`).slice(0, 3)}
				</Text>
			</View>
		</Link>
	);
}
