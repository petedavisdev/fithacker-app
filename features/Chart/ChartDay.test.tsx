import * as React from 'react';
import renderer from 'react-test-renderer';

import { ChartDay } from './ChartDay';
jest.mock('react-i18next', () => ({
	useTranslation: () => ({ t: (str: string) => str }),
}));

it(`renders correctly`, () => {
	const tree = renderer
		.create(<ChartDay date="2022-01-01" exercises={['🚶', '🏃‍♀️']} />)
		.toJSON();

	expect(tree).toMatchSnapshot();
});
