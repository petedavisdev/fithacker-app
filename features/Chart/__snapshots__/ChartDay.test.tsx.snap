// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Text
  disabled={false}
  href="/?date=2022-01-01"
  onPress={[Function]}
  role="link"
>
  <View
    className="justify-end items-center h-96 gap-2"
  >
    <View
      className="relative"
    >
      <Text
        className="text-yellow-500 text-4xl"
      >
        🚶
      </Text>
    </View>
    <View
      className="relative"
    >
      <Text
        className="text-yellow-500 text-4xl"
      >
        🏃‍♀️
      </Text>
    </View>
    <View
      className="h-[2px] w-12 bg-yellow-500 shadow shadow-yellow-500"
    />
    <Text
      className="font-mono leading pb-0.5 border-b text-yellow-500 border-yellow-700"
    >
      _da
    </Text>
  </View>
</Text>
`;
